### PR TITLE
Tasks: moves show completed toggle to sidebar

### PR DESCRIPTION
**Pull request by AbleLincoln**
Changes input to a toggle.
Also, `showCompleted` is in redux store

### CODE REVIEW

File: src/components/Tasks/Sidebar.tsx [hunk 1]
```diff
@@ -61,12 +80,16 @@ export default React.memo(function Sidebar(props: { tasks: TaskType[] }) {
...
   return (
-    <List dense>
+    <List>
       <SidebarListItem name={null} primaryText="All" icon={<InboxIcon />} amount={tasks.length} />
       <SidebarListItem name="today" primaryText="Due today" icon={<TodayIcon />} amount={amountDueToday} />
 
       <ListSubheader>Tags</ListSubheader>
       {tagsList}
+
+      <Divider style={{ marginTop: '1em', marginBottom: '1em' }} />
```
Comment (tasn) [inline src/components/Tasks/Sidebar.tsx]: You can just use 'marginVertical'

Comment (AbleLincoln) [reply src/components/Tasks/Sidebar.tsx]: oh cool I had no idea. Is that a React feature? or MUI?

Comment (tasn) [reply src/components/Tasks/Sidebar.tsx]: Please ignore me, it's a react-native thing, not a reactjs thing.

**Comment by AbleLincoln** [comment]
I liked it better there too, but I'm kind of running out of space. I have to fit sorting, searching, and these toggles somewhere. When I tried to place the sorting on the top where the toggle used to be, it looked weird. That's why I am moving the toggle in the first place.

That screen shot is great, and I've tried to refer to it. Our design is just a little different with the nav tabs and paper and everything. I wonder, do you think I should make some kind of toolbar like the example? Or I could do a three dot menu, like on Tasks.org?

**Comment by AbleLincoln** [comment]
Also, I see what you're saying about settings and pages. My main reason I was using the settings object was because I wanted some kind of persistent state. From a UX perspective it's frustrating that every time I go back Tasks list (whether from a different page, or from editing) it's in some default state, rather than the way I left it.

**Comment by tasn** [comment]
> I liked it better there too, but I'm kind of running out of space. I have to fit sorting, searching, and these toggles somewhere. When I tried to place the sorting on the top where the toggle used to be, it looked weird. That's why I am moving the toggle in the first place.
> 
> That screen shot is great, and I've tried to refer to it. Our design is just a little different with the nav tabs and paper and everything. I wonder, do you think I should make some kind of toolbar like the example? Or I could do a three dot menu, like on Tasks.org?

I'd say: very similar to Tasks.org. A search icon directly there, maybe also sort (though probably not) and then the rest in the dot menu. Sort, show completed and etc aren't *that* common to normal workflow.

**Comment by tasn** [comment]
> Also, I see what you're saying about settings and pages. My main reason I was using the settings object was because I wanted some kind of persistent state. From a UX perspective it's frustrating that every time I go back Tasks list (whether from a different page, or from editing) it's in some default state, rather than the way I left it.

You can also have it in the navigation context, or even the navigation path itself. I don't know what's better so don't worry about it. Maybe sometime in the future I (or you) will change it. I just wanted to share my thoughts on this in case it affects your future work on this.

As for the "every time I go back", it should work just fine when you use the navigation context, that's actually what it's for. Think of it as kind of a query string, so the browser (and the app) already know how to deal with it.

**Comment by AbleLincoln** [comment]
Try this. I've moved the option to a drop down menu

**Comment by AbleLincoln** [comment]
> You can also have it in the navigation context, or even the navigation path itself. 

Interesting, then the "state" is controlled by the router essentially?

I'll keep this in mind as I develop, and we discuss refactoring to this after the feature is more complete.

**Comment by tasn** [comment]
It's "state" in a similar way to how to how the url is state (though admittedly a bit more). Yeah, though, the nice thing about it is that it maintains the stackable nature of state.

File: src/components/Tasks/Sidebar.tsx [hunk 2]
```diff
@@ -61,7 +61,7 @@ export default React.memo(function Sidebar(props: { tasks: TaskType[] }) {
   ));
 
   return (
-    <List dense>
```
Comment (tasn) [inline src/components/Tasks/Sidebar.tsx]: This is an unrelated change to this PR.

File: src/components/Tasks/TaskList.tsx [hunk 3]
```diff
@@ -37,10 +35,10 @@ interface PropsType {
 }
 
 export default function TaskList(props: PropsType) {
-  const [showCompleted, setShowCompleted] = React.useState(false);
   const settings = useSelector((state: StoreState) => state.settings.taskSettings);
   const { filterBy } = settings;
   const theme = useTheme();
+  const [showCompleted, setShowCompleted] = React.useState(false);
```
Comment (tasn) [inline src/components/Tasks/TaskList.tsx]: Please move it back to here it was (look at the diff of the change as a whole). It's just an unrelated change now.

File: src/components/Tasks/Toolbar.tsx [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+import QuickAdd from './QuickAdd';
+
+import { PimType } from '../../pim-types';
+
+interface PropsType {
+  defaultCollection: EteSync.CollectionInfo;
+  onItemSave: (item: PimType, journalUid: string, originalItem?: PimType) => Promise<void>;
+  showCompleted: boolean;
+  setShowCompleted: React.Dispatch<React.SetStateAction<boolean>>;
```
Comment (tasn) [inline src/components/Tasks/Toolbar.tsx]: This is not the correct type (though it seems to work for you). The correct type is `(completed: boolean) => void`.

File: src/components/Tasks/Toolbar.tsx [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+export default function Toolbar(props: PropsType) {
+  const { defaultCollection, onItemSave, showCompleted, setShowCompleted } = props;
+
+  const [anchorEl, setAnchorEl] = React.useState<null | HTMLElement>(null);
```
Comment (tasn) [inline src/components/Tasks/Toolbar.tsx]: You are probably looking for https://reactjs.org/docs/refs-and-the-dom.html#creating-refs rather than manually setting the ref like this. Though I see that's the recommended way for MUI, so I guess keep it this way. Though :nauseated_face:

File: src/components/Tasks/Toolbar.tsx [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+          aria-haspopup="true"
+          onClick={handleClick}
+        >
+          <MoreVertIcon />
+        </IconButton>
+        <Menu
+          id="simple-menu"
+          anchorEl={anchorEl}
+          keepMounted
+          open={Boolean(anchorEl)}
```
Comment (tasn) [inline src/components/Tasks/Toolbar.tsx]: We use `!!` for boolean coercion in the rest of the code, so if you can change here too.

File: src/components/Tasks/Toolbar.tsx [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    setAnchorEl(event.currentTarget);
+  };
+
+  const handleClose = () => {
+    setAnchorEl(null);
+  };
+
+  return (
+    <div style={{ display: 'flex', justifyContent: 'space-between' }}>
+      {defaultCollection && <QuickAdd onSubmit={onItemSave} defaultCollection={defaultCollection} />}
```
Comment (tasn) [inline src/components/Tasks/Toolbar.tsx]: I missed it in a previous change, but it looks like there's a hardcoded `25px` right margin in QuickAdd. It should be removed from there entirely (as this is not an internal property of QuickAdd) and instead added here.
The spacing is also wrong because:
1. It's too spacious.
2. It's using pixel sizing rather than `em`.
It should be changed to `0.75em` which I checked and looks good locally.

File: src/components/Tasks/Toolbar.tsx [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+          id="simple-menu"
+          anchorEl={anchorEl}
+          keepMounted
+          open={Boolean(anchorEl)}
+          onClose={handleClose}
+        >
+          <MenuItem>
+            <FormControlLabel
+              label="Show completed"
+              control={<Switch checked={showCompleted} onChange={(_e, checked) => setShowCompleted(checked)} />}
```
Comment (tasn) [inline src/components/Tasks/Toolbar.tsx]: I think that if it's a switch, the switch should be on the left and the label on the right, and if it's checkbox the way it is now. What do you think? It just looks odd to me like this.

File: src/components/Tasks/Toolbar.tsx (hunk 5 above)
Comment (AbleLincoln) [reply src/components/Tasks/Toolbar.tsx]: Yeah I just took it directly off their docs 😬

File: src/components/Tasks/Toolbar.tsx (hunk 8 above)
Comment (AbleLincoln) [reply src/components/Tasks/Toolbar.tsx]: Do you mean the other way around? I think the switch is currently on the left

Comment (tasn) [reply src/components/Tasks/Toolbar.tsx]: Yeah, the other way around. Sorry, I meant switch on the right. :)

Comment (AbleLincoln) [reply src/components/Tasks/Toolbar.tsx]: I put it that way and you're right, looks much better.

Comment (tasn) [reply src/components/Tasks/Toolbar.tsx]: Great. :)

File: src/components/Tasks/Toolbar.tsx [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+    setAnchorEl(event.currentTarget);
+  };
+
+  const handleClose = () => {
+    setAnchorEl(null);
+  };
+
+  return (
+    <div style={{ display: 'flex', justifyContent: 'space-between' }}>
+      {defaultCollection && <QuickAdd style={{ flexGrow: 1, marginRight: '0.75em' }} onSubmit={onItemSave} defaultCollection={defaultCollection} />}
```
Comment (tasn) [inline src/components/Tasks/Toolbar.tsx]: Ah, nice catch moving the flexGrow too, I missed that one. :)
